### DISCUSSION TITLE
Mako Sidecar Creation - new XMP fields

### DISCUSSION THREAD

**Discussion by acronomic** [RhetTbull/osxphotos#1606] [General]
Hi,  

  I'm looking to build out a sidecar that will be readable by Neofinder. I'm not familiar with Mako and wondering how to form some of the <%def> block names. eg I looked at an xmp file generated from neofinder and it has a <photoshop:TransmissionReference> tag, would this translate to <%def name="photoshop_transmission_reference(jobId)"> or something else? Is there a way to find out all the legally formed names or to test if one is present? 

How would this work for custom fields in the xmp like the following:

         <expressionmedia:UserFields>
            <rdf:Seq>
               <rdf:li>pDuration=nxCTpDuration</rdf:li>
            </rdf:Seq>
         </expressionmedia:UserFields>

Thank you.

**Comment by RhetTbull** [comment]
I am not familiar with neofinder nor with these XMP fields you've mentioned. Adding custom fields isn't hard and the [example](https://github.com/RhetTbull/osxphotos/blob/main/examples/custom_sidecar_xmp.mako) should get you started. To use Mako you need to understand a bit of python (the control logic is a subset of python) and you need to understand the field you're trying to add. For example, `photoshop:TransmissionReference` appears to be [some sort of job ID](https://community.adobe.com/t5/lightroom-classic-discussions/job-identifier-tag-in-the-metadata-panel/td-p/11702781). Where are you getting this value from? I'm not aware of any value in the Photos database that would equate to this? It sounds to me like it's a way to associate photos from a particular job together but this isn't something Photos stores so where do you plan to get this value from when you're writing the custom XMP file during export?

> How would this work for custom fields in the xmp like the following:

The Mako file is just an XMP file (or an HTML file or a JSON file or anything you want it do be). The Mako control logic (tags beginning with `<%def` and `%` as well as execution blocks that look like `${...}`) is all replaced at render time with the new values but anything else in the file remains. So to add the example to your file, you'd just include the following in template file:

```xmp
    <expressionmedia:UserFields>
        <rdf:Seq>
           <rdf:li>pDuration=nxCTpDuration</rdf:li>
        </rdf:Seq>
     </expressionmedia:UserFields>
```

I'm happy to provide some pointers if I better understand what fields you're trying to include and where that data should come from but I don't use/know neofinder so can't comment on specifics of it's desired format.

**Reply by RhetTbull** [reply]
The other question is what is your use case? In other words, what are you trying to accomplish? Does NeoFinder recognize the regular XMP sidecars generated by OSXPhotos? They are standard sidecars and should be compatible with any app that reads XMP (e.g. Lightroom, DigiKam, etc.).  Why are the additional fields necessary? Can they be added by NeoFinder after import into NeoFinder? If you can provide a better idea of exactly what you're trying to do and what the current limitations are, that would be helpful.

**Reply by acronomic** [reply]
Thank you so much for the quick reply.  Neofinder is a DAM like lightroom.  I use it to manage my photos and it allows for sidecar XMP metadata.  I used the custom_sidecar_xmp.mako file to test compatibility and it generated xmps that were readable in Neofinder.  

I believe I was confusing the fields from Neofinder with the fields from osxphotos.   `photoshop:TransmissionReference` is something I found when inspecting a sidecar Neofinder generated.  What I would like to do is take values from osxphotos and format them into the xmp fields Neofinder expects.  I understand the syntax that I need to generate for the XMP, but I guess where I'm lost is how to get the values out of osxphotos.  

In the following example:

```mako
<%def name="dc_datecreated(date)">
    % if date is not None:
        <photoshop:DateCreated>${date.isoformat()}</photoshop:DateCreated>
    % endif
</%def>
```

Where does the variable name dc_datecreated come from? Is there a reference of all these that are in osxphotos?  I'm sure I'm just missing something small.

An example of what I'd like to do export the OSXPHOTOS Place information to the XMP Source field `<photoshop:Source>` .

Thanks again.

**Comment by RhetTbull** [comment]
> Where does the variable name dc_datecreated come from?

`dc_datecreated(date)` is a function defined in Mako that gets passed the date. It gets called later in the template file to render the date:

`${dc_datecreated(photo.date)}`

The `photo` variable is a [PhotoInfo](https://rhettbull.github.io/osxphotos/API_README.html#photoinfo) object that provides access to all the data OSXPhotos has for the photo.

Reference the help for the sidecars template option:


--sidecar-template <MAKO_TEMPLATE_FILESIDECAR_FILENAME_TEMPLATEOPTIONS>
:

Create a custom sidecar file for each photo exported with user provided Makotemplate (MAKO_TEMPLATE_FILE). MAKO_TEMPLATE_FILE must be a valid Mako template (see [https://www.makotemplates.org/](https://www.makotemplates.org/)). The template will passed the following variables: photo (PhotoInfo object for the photo being exported), sidecar_path (pathlib.Path object for the path to the sidecar being written), and photo_path (pathlib.Path object for the path to the exported photo. SIDECAR_FILENAME_TEMPLATE must be a valid template string (see Templating System in help) which will be rendered to generate the filename of the sidecar file. The {filepath} template variable may be used in the SIDECAR_FILENAME_TEMPLATE to refer to the filename of the photo being exported. OPTIONS is a comma-separated list of strings providing additional options to the template. Valid options are: write_skipped, strip_whitespace, strip_lines, skip_zero, catch_errors, none. write_skipped will cause the sidecar file to be written even if the photo is skipped during export. If write_skipped is not passed as an option, the sidecar file will not be written if the photo is skipped during export. strip_whitespace and strip_lines indicate whether or not to strip whitespace and blank lines, respectively, from the resulting sidecar file. skip_zero causes the sidecar file to be skipped if the rendered template is zero-length. catch_errors causes errors in the template to be caught and logged but not raised. Without catch_errors, osxphotos will abort the export if an error occurs in the template. For example, to create a sidecar file with extension .xmp using a template file named 'sidecar.mako' and write a sidecar for skipped photos and strip blank lines but not whitespace: --sidecar-template sidecar.mako '{filepath}.xmp' write_skipped,strip_lines. To do the same but to drop the photo extension from the sidecar filename: --sidecar-template sidecar.mako '{filepath.parent}/{filepath.stem}.xmp' write_skipped,strip_lines. If you are not passing any options, you must pass 'none' as the last argument to --sidecar-template: --sidecar-template sidecar.mako '{filepath}.xmp' none. For an example Mako file see <https://raw.githubusercontent.com/RhetTbull/osxphotos/main/examples/custom_sidecar.mako>](https://raw.githubusercontent.com/RhetTbull/osxphotos/main/examples/custom_sidecar.mako)
